pytorch_lightning/metrics/retrieval/retrieval_metric.py

```diff
@@ -0,0 +1,140 @@
+from abc import ABC, abstractmethod
+import torch
+from typing import Optional, Callable, Any
+
+from pytorch_lightning.metrics import Metric
+from pytorch_lightning.metrics.utils import get_group_indexes
+#: get_group_indexes is used to group predictions belonging to the same query
+
+IGNORE_IDX = -100
+
+
+class RetrievalMetric(Metric, ABC):
+    r"""
+    Works with binary data. Accepts integer or float predictions from a model output.
+
+    Forward accepts
+    - ``indexes`` (long tensor): ``(N, ...)``
+    - ``preds`` (float or int tensor): ``(N, ...)``
+    - ``target`` (long or bool tensor): ``(N, ...)``
+
+    `indexes`, `preds` and `target` must have the same dimension and will be flatten
+    to single dimension once provided.
+
+    `indexes` indicate to which query a prediction belongs.
+    Predictions will be first grouped by indexes. Then the
+    real metric, defined by overriding the `_metric` method,
+    will be computed as the mean of the scores over each query.
+
+    Args:
+        compute_on_step:
+            Forward only calls ``update()`` and return None if this is set to False. default: True
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step. default: False
+        process_group:
+            Specify the process group on which synchronization is called. default: None (which selects
+            the entire world)
+        dist_sync_fn:
+            Callback that performs the allgather operation on the metric state. When `None`, DDP
+            will be used to perform the allgather. default: None
+        query_without_relevant_docs:
```

Comment:
please reorder args in docstring so they correspond to order they appear in `__init__`